imap_processing/hi/l1a/science_direct_event.py

```diff
@@ -0,0 +1,288 @@
+    # but at present there is no way to query what tick
+    # duration is in use.  I am attempting to fix that
+    # issue. It looks like the raw units of tick
+    # duration are very close to 1 us increments
+    # (3999 corresponds to 4 ms).
+    # TODO: ask what value should this be for now.
+    looked_up_duration_of_tick = 3999  # ??
+    time_in_ns = (
+        met_seconds * 1e9
+        + met_subseconds * 1e6
+        + de_tag * 1e6
+        + looked_up_duration_of_tick * 1e3
+    )
+    met_datetime = launch_time + np.timedelta64(int(time_in_ns), "ns")
+
+    return met_datetime
+
+
+def parse_direct_event(event_data: str):
+    """Parse event data.
+
+    IMAP-Hi direct event data information is stored in
+    48-bits as follow:
+        Read first two bits (start_bitmask_data) to find
+        out which type of event it is. start_bitmask_data value mapping:
+            1 - A
+            2 - B
+            3 - C
+            0 - META
+        If it's a metaevent:
+            Read 48-bits into 2, 4, 10, 32 bits. Each of these breaks
+            down as:
+                start_bitmask_data - 2 bits (tA=1, tB=2, tC1=3, META=0)
+                ESA step - 4 bits
+                integer millisecond of MET(subseconds) - 10 bits
+                integer MET(seconds) - 32 bits
+        If it's not a metaevent:
+            Read 48-bits into 2, 10, 10, 10, 16 bits. Each of these breaks
+            down as:
+                start_bitmask_data - 2 bits (tA=1, tB=2, tC1=3, META=0)
+                tof_1 - 10 bit counter
+                tof_2 - 10 bit counter
+                tof_3 - 10 bit counter
+                de_tag - 16 bits
+
+    There are at most total of 665 of 48-bits in each data packet.
+    This data packet is of variable length. If there is one event, then
+    DE_TOF will contain 48-bits. If there are 665 events, then
+    DE_TOF will contain 665 x 48-bits. If there is no event, then
+    DE_TOF will contain 0-bits.
+
+    Per ESA, there should be two data packets. First one will begin with
+    metaevent followed by direct events data. Second one will begin with
+    direct event data only. If there is no event record for certain ESA step,
+    then as mentioned above, first packet will contain metaevent in DE_TOF
+    information and second packet will contain 0-bits in DE_TOF. In general,
+    every two packets will look like this.
+        first packet = [
+            (start_bitmask_data, ESA step, int millisecond of MET, int MET),
+            (start_bitmask_data, tof_1, tof_2, tof_3, de_tag),
+            .....
+        ]
+        second packet = [
+            (start_bitmask_data, tof_1, tof_2, tof_3, de_tag),
+            .....
+        ]
+
+    In direct event data, if no hit is registered, the tof_x field in
+    the DE to a value of negative one. However, since the field is described as a
+    "10-bit unsigned counter," it cannot actually store negative numbers.
```

Comment:
Interesting. I used a fill value for events with no hit. We should be consistent with this. Let's discuss at some point, but this is ok for now.

Reply:
IMAP-Hi still want to use 1023 because this will indicate that the detect was not hit. They use that value to annotate single, double or triple coincidence in l1b.